compiler/resolution/lowerForalls.cpp

```diff
@@ -873,6 +893,89 @@ static void expandTopLevel(ExpandVisitor* outerVis,
 }
 
 
+/////////// reorder shadow variables ///////////
+
+// "Proper" shadow variables i.e. not TPVs should go first,
+// except those for reduce intents.
+static bool shouldGoFirst(ShadowVarSymbol* sv) {
+  switch (sv->intent)
+  {
+  case TFI_DEFAULT:
+  case TFI_CONST:
+  case TFI_IN_OUTERVAR:
+  case TFI_IN:
+  case TFI_CONST_IN:
+  case TFI_REF:
+  case TFI_CONST_REF:
+    return true;
+
+  case TFI_REDUCE:
+  case TFI_REDUCE_OP:
+  case TFI_TASK_PRIVATE:
+    return false;
+  }
+
+  return false; // dummy
+}
+
+//
+// We want all proper shadow variables to be ahead of task-private variables.
+// Explanation: (a) The latter may reference the former - in TPVs initBlocks.
+// (b) We compute+apply SymbolMaps incrementally. (c) When we map a TPV,
+// the proper shadow variables that it references must already be in the map.
+//
+// We do not hoist reduce-intent shadow variables. This is because they
+// should not reference proper shadow variables (right?). So mapping
+// is not an issue. The issue is that we want to preserve the order of
+// initialization of reduction shadow variables w.r.t. task-private variables
+// - because initialization actions may have visible side-effects.
+//
+static void reorderShadowVsTaskPrivateVars(ForallStmt* fs) {
+  AList& svars = fs->shadowVariables();
+  if (svars.empty()) return; // nothing to reorder
+
+  Expr* currDef = svars.head;
+  Expr* lastReorderedDef = NULL;
+
+  do {
```

Comment:
I think this loop would probably be easier to understand if you wrote it with 2 lists, where you put the "should go first" ones in one list, and the others in another list, and then output the list elements in the right order. I'm not sure what's gonig on with "not reorderable".